pkg/export/secrets/watch.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// WatchSPConfig configures access to the Kubernetes API server.
+// TODO(TheSpiritXIII): https://github.com/GoogleCloudPlatform/prometheus-engine/issues/867
+type WatchSPConfig struct {
+	ClientConfig
+}
+
+func (c *WatchSPConfig) newProvider(ctx context.Context, opts ProviderOptions) (*watchProvider, error) {
+	client, err := c.ClientConfig.client()
+	if err != nil {
+		return nil, err
+	}
+	return newWatchProvider(ctx, opts.Logger, client)
+}
+
+type watcher struct {
+	mu       sync.Mutex
+	w        watch.Interface
+	refCount uint
+	s        *corev1.Secret
+}
+
+func newWatcher(ctx context.Context, logger log.Logger, client kubernetes.Interface, config *KubernetesSecretConfig) (*watcher, error) {
+	val := &watcher{
+		refCount: 1,
+	}
+
+	var err error
+	val.w, err = client.CoreV1().Secrets(config.Namespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector:       fields.OneTermEqualSelector(metav1.ObjectNameField, config.Name).String(),
+		AllowWatchBookmarks: false,
+	})
+	if err != nil {
+		return val, fmt.Errorf("unable to watch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+
+	// We could wait for the first watch event, but it doesn't notify us if the resource doesn't exist.
+	val.s, err = client.CoreV1().Secrets(config.Namespace).Get(ctx, config.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return val, fmt.Errorf("unable to fetch secret %s/%s: %w", config.Namespace, config.Name, err)
+	}
+	go func() {
+		for {
+		channelLoop:
+			for {
+				select {
+				case e, ok := <-val.w.ResultChan():
+					if !ok {
+						break channelLoop
+					}
+					val.update(logger, e)
+				case <-ctx.Done():
+					// Run within separate function so we can use defer.
+					func() {
+						val.mu.Lock()
+						defer val.mu.Unlock()
+						val.w.Stop()
+					}()
+					return
+				}
+			}
+
+			// Run within separate function so we can use defer.
+			finished := func() bool {
```

Comment:
question: What's this routine doing, qualitatively? Some cleanup? Restarting the watch? Why?

suggestion: And can you add comments to explain?